File: wetty-chat-mobile/src/api/invites.ts
import type { AxiosResponse } from 'axios';
import apiClient from './client';
import type { GroupInfoResponse } from './group';
import type { MessageResponse } from './messages';

export type InviteType = 'generic' | 'targeted' | 'membership';

export interface InviteInfoResponse {
  id: string;
  code: string;
  chatId: string;
  inviteType: InviteType;
  creatorUid: number | null;
  targetUid: number | null;
  requiredChatId: string | null;
  createdAt: string;
  expiresAt: string | null;
  revokedAt: string | null;
  usedAt: string | null;
}

export interface InvitePreviewResponse {
  invite: InviteInfoResponse;
  chat: GroupInfoResponse;
  alreadyMember: boolean;
}

export interface RedeemInviteBody {
  code: string;
}

export interface RedeemInviteResponse {
  chat: GroupInfoResponse;
}

export interface ListInvitesResponse {
  invites: InviteInfoResponse[];
}

export interface CreateInviteBody {
  chatId: string;
  inviteType: InviteType;
  targetUid?: number;
  requiredChatId?: string;
  expiresAt?: string | null;
}

export interface SendInviteMessageBody {
  sourceChatId: string;
  destinationChatId: string;
  inviteId?: string;
  expiresAt?: string | null;
  clientGeneratedId: string;
}

export interface SendInviteMessageResponse {
  invite: InviteInfoResponse;
  message: MessageResponse;
}

export function getInvitePreview(inviteCode: string): Promise<AxiosResponse<InvitePreviewResponse>> {
  return apiClient.get('/invites/invite', { params: { inviteCode } });
}

export function redeemInvite(body: RedeemInviteBody): Promise<AxiosResponse<RedeemInviteResponse>> {
  return apiClient.post('/invites/redeem', body);
}

export function getInvites(
  params: {
    groupId?: string;
    limit?: number;
  } = {},
): Promise<AxiosResponse<ListInvitesResponse>> {
  return apiClient.get('/invites', { params });
}

export function createInvite(body: CreateInviteBody): Promise<AxiosResponse<InviteInfoResponse>> {
  return apiClient.post('/invites', body);
}

export function sendInviteMessage(body: SendInviteMessageBody): Promise<AxiosResponse<SendInviteMessageResponse>> {
  return apiClient.post('/invites/send', body);
}

export function deleteInvite(inviteId: string): Promise<AxiosResponse<void>> {
  return apiClient.delete(`/invites/invite/${inviteId}`);
}

File: backend/src/dto/invites.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use utoipa::ToSchema;

use crate::{
    dto::{groups::GroupInfoResponse, messages::MessageResponse},
    models::InviteType,
};

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct InviteResponse {
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub id: i64,
    pub code: String,
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub chat_id: i64,
    pub invite_type: InviteType,
    pub creator_uid: Option<i32>,
    pub target_uid: Option<i32>,
    #[serde(with = "crate::serde_i64_string::opt")]
    #[schema(value_type = Option<String>)]
    pub required_chat_id: Option<i64>,
    pub created_at: DateTime<Utc>,
    pub expires_at: Option<DateTime<Utc>>,
    pub revoked_at: Option<DateTime<Utc>>,
    pub used_at: Option<DateTime<Utc>>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ListInvitesResponse {
    pub invites: Vec<InviteResponse>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct InvitePreviewResponse {
    pub invite: InviteResponse,
    pub chat: GroupInfoResponse,
    pub already_member: bool,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct RedeemInviteResponse {
    pub chat: GroupInfoResponse,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SendInviteMessageResponse {
    pub invite: InviteResponse,
    pub message: MessageResponse,
}
